### DISCUSSION TITLE
Solution for apps reverting their icons when updating

### DISCUSSION THREAD

**Discussion by chriskalos** [elrumo/macOS_Big_Sur_icons_replacements#1111] [Ideas]
Hi all!

I will try to make a script that one can run every time an app updates and loses its custom icon, to customize it again without going into its Info and dragging the icon into that little window all over again.

My idea is:
- User sets a folder for their custom icons
- Tells the script what icon to use for each app (probably through a config file)
- Runs script when they want to batch-change all the icons in one go
- Script restarts Finder to reload icons

### **EDIT**:
It is actually way more difficult to make the script than I thought. At this point, the end user will have to modify the script so much to get it to work for their machine, that they may as well not use it at all.
However, I can give y'all the template for creating it yourself! This is what it looks like, using some code from [this guide](https://www.sethvargo.com/replace-icons-osx).

```
#!/bin/zsh
osascript -e 'quit app "[Application]"' &&
cp "[directory]/[custom icon].icns" "/Applications/[Application].app/Contents/Resources/[app icon].icns"
touch "/Applications/[Application].app"

# The user can insert more sets of commands along those lines for more applications.

# At the end, the script will reset Finder and Dock to accept the changes.
sudo killall Finder && sudo killall Dock
```

The `osascript -e 'quit app "[]"'` command in the beginning is an AppleScript that tells the system to safely quit the app whose icon is about to be changed. The `&&` prevents the script from going on until the application has quit successfully.

The `cp` command copies the custom icon data to the internal icon of the application in question.
**IMPORTANT NOTE:** Some apps have quirks. For example, Microsoft Edge's app icon is called app.icns in the /Contents/Resources/ directory, but OneDrive's is called OneDrive.icns. Make sure to replace the file names and directories in the `cp` command correctly. As far as I can tell, the filename of your custom icon doesn't matter, as long as you replace the correct file, it will be renamed to it.

The `touch` command, as [Seth Vargo](https://www.sethvargo.com/replace-icons-osx) explains, is an effective way to get macOS to reset the icon cache of that app.

At the end, we `sudo killall Finder && sudo killall Dock` to restart the Finder and Dock processes and thus reload all the icons.

**EXAMPLE**:
```
#!/bin/zsh
osascript -e 'quit app "Microsoft Edge"' &&
cp "/Users/blorp/Custom Icons/Microsoft Edge.icns" "/Applications/Microsoft Edge.app/Contents/Resources/app.icns"
touch "/Applications/Microsoft Edge.app"
sudo killall Finder && sudo killall Dock
```
The above script will replace the icon of Microsoft Edge to one that I have placed in the "Custom Icons" folder of the user directory of my imaginary friend "blorp".

To get such a script to run, you must write it in your text editor of choice and save it as a .sh file, and then do `chmod +x [file]` (example: `chmod +x replace.sh`) to make it executable. After that, you can run it by using `./[file]` (example: `./replace.sh`) and it'll do all the things you have told it to do.

**Comment by NewsGuyTor** [comment]
You should of course use [Pictogram](https://pictogramapp.com/) for non App Store apps, but this is great for App Store apps.

**Reply by toioski** [reply]
@gl-Gho I don't understand what you use `sudo -i` for. Could you elaborate a bit more please?
